Pathfinding: pair-based liquidity hints and sync from mission control

bitromortac opened the pull request:
This PR synchronizes mission control data from the lnd node and populates liquidity hints, where liquidity hints are now keyed by node pubkey pairs. This way we have an easy way to benefit from the existing data if lnd is used with other tools. Liquidity hints are extended with timestamps for each individual information (can_send/cannot_send in both directions).

Todo:
- [ ] deal with number of forwarding trials when mission control data is loaded

bitromortac committed:
data_types: add key pair

bitromortac committed:
data_types: move shortchanid and nodeid

bitromortac committed:
liquidityhints: switch to pair based keying

bitromortac committed:
network: use max capacity per node pair

bitromortac committed:
liquidityhints: extend with timestamps

bitromortac committed:
liqudityhints: extend with mission control data

bitromortac committed:
liquidityhints: synchronize mc data conditionally

bitromortac committed:
lndmanage: improve startup messages

3nprob commented on lndmanage/lib/liquidityhints.py:
```diff
@@ -17,12 +21,22 @@
+        return bool(self.amount) and bool(self.timestamp)
+
+    def __gt__(self, other):
+        return self and self.amount > other.amount
```
I guess this should check  for truthiness of `other` as well?
Like `return self and (not other or self.amount > other.amount)`

Also it looks like since `bool(x) == False` if `x.amount == 0`, for any `x: x.amount == 0` and any `y`, `!(x < y) && !(x > y)`.
So this would break the consistency assumptions for these operators.

Maybe not a practical issue here, just bringing it up for consideration.

bitromortac replied:
good comment, I changed it to also check for bool(other), it also simplifies the comparison, before I always checked truthiness explicitly

3nprob commented on lndmanage/lib/liquidityhints.py:
```diff
@@ -17,12 +21,22 @@
-    pass
+@dataclass
+class AmountHistory:
+    amount: int = 0  # msat
```
Would it make sense to initialize to `None`, in order to distinguish between undefined vs actual 0?

bitromortac replied:
Yes, that makes more sense! I changed it.

3nprob commented on lndmanage/lib/data_types.py:
```diff
@@ -46,3 +46,18 @@ class NodeProperties:
     sent_received_per_week: int
+
+
+class KeyPair(tuple):
```
Just noting that this may be a confusing naming due to the more common specific meaning of `key pair`.

Suggestions: `NodePair`, `PubkeyTuple`, `NodeIDPair`, `NodeIDTuple`.

bitromortac replied:
Had that thought as well, should have done that from the beginning, landed in rebase hell :grin:

3nprob commented on lndmanage/lib/data_types.py:
```diff
@@ -46,3 +46,18 @@ class NodeProperties:
+class KeyPair(tuple):
+    """KeyPair represents a node pair mapped to a fixed order."""
+    def __new__(cls, keys):
+        seq = (keys[0], keys[1]) if keys[0] < keys[1] else (keys[1], keys[0])
```
Suggestion: Make it case-insensitive:
`         seq = (keys[0], keys[1]) if keys[0].casefold() < keys[1].casefold() else (keys[1], keys[0])`.


This will also make it raise when keys are not instances of (subclasses of) str, which I guess is desirable.

3nprob commented on lndmanage/lib/liquidityhints.py:
```diff
@@ -257,26 +339,33 @@ def penalty(self, node_from: NodeID, node_to: NodeID, edge: Dict, amount_msat: i
         """
         # we assume that our node can always route:
-        if self.source_node in [node_from, ]:
+        if self.source_node in [
```
Just curious: Why an inclusion check rather than just `if self.source_node == node_from`?

bitromortac replied:
This is a leftover from when I was experimenting with pathfinding (before I checked also `node_to` in the list). Thanks for making me aware.

3nprob commented on lndmanage/lib/liquidityhints.py:
```diff
@@ -333,3 +432,29 @@ def __repr__(self):
+    def extend_with_mission_control(self, mc_pairs):
+        logger.info("> Syncing mission control data.")
+        for pair in mc_pairs:
+            node_from = pair.node_from.hex()
```
`node_from = NodeID(pair.node_from.hex())`?

bitromortac replied:
yes, I should be more consistent with typing for node pubkeys

3nprob commented on lndmanage/lib/rating.py:
```diff
@@ -52,13 +53,16 @@ def node_to_node_weight(self, u, v, e, amt_msat):
+
         # we don't send if the channel cannot carry the payment
-        if amt_msat // 1000 > channel_info["capacity"]:
+        if amt_msat // 1000 > capacity:
```
This can miss cases where fractional sats still make it over. 

```
if amt_msat > capacity * 1000:
```

bitromortac replied:
right, thanks!

bitromortac committed:
network: add node pair to channels

bitromortac committed:
liquidityhints: remove hint timestamp

bitromortac committed:
liquidityhints: remove judge for backward dir

Take a step back and remove this assumption.

bitromortac committed:
liquidityhints: change direction

This is the intended direction syntax, where the comarison operator
indicates the direction from where we want to send: from_node > to_node.

Caution: invalidates preexisting liquidity hints

bitromortac commented:
@3nprob thank you for the comments, that's really great :+1:! I tried to incorporate your thoughts and found some bugs on the way as well. I'll leave this open still for a few days if you may want to give further feedback.

3nprob commented:
This looks great!

3nprob commented on lndmanage/lib/liquidityhints.py:
```diff
@@ -17,12 +21,22 @@
-    pass
+@dataclass
+class AmountHistory:
+    amount: int = None  # msat
```
Minor nit: Could be renamed to `amount_msat`

bitromortac committed:
hints: rename to msat

bitromortac committed:
hints: warn about old pickle files
